libqtile/widget/pulse_volume.py

```diff
@@ -0,0 +1,253 @@
+            self.handle
+        )
+        self.wait_for_operation(op)
+        if not self._subscribed:
+            self.subscribe_to_sink_events()
+
+    def on_sink_info(self, sink, eol):
+        if eol:  # dont operate on sink in case its an eol callback
+            return
+        name = ffi.string(sink.name).decode('utf-8')
+        if name == self.default_sink_name:
+            self.default_sink = {
+                'name': name,
+                'description': ffi.string(sink.description).decode('utf-8'),
+                'index': int(sink.index),
+                'base_volume': sink.base_volume,
+                'muted': bool(sink.mute),
+                'channels': sink.volume.channels,
+                'values': list(sink.volume.values),
+            }
+            self.update()
+
+    def subscribe_to_sink_events(self):
+        op = lib.pa_context_subscribe(
+            self.context,
+            lib.PA_SUBSCRIPTION_MASK_SINK,
+            ffi.NULL,
+            ffi.NULL,
+        )
+        self.wait_for_operation(op)
+        lib.pa_context_set_subscribe_callback(
+            self.context,
+            lib.qtile_on_sink_update,
+            self.handle
+        )
+        self._subscribed = True
+
+    def on_sink_update(self, event_type, sink_index):
+        self.timeout_add(0.1, self.get_sinks)
+
+    def wait_for_operation(self, op, unref=True):
+        state = lib.pa_operation_get_state(op)
+        while state == lib.PA_OPERATION_RUNNING:
+            lib.pa_mainloop_iterate(self.loop, 0, ffi.NULL)
+            state = lib.pa_operation_get_state(op)
+
+    def change_volume(self, volume):
+        """
+        order pulseaudio to apply new volume
+        """
+        # store new volume to "speed up" widget update so that we don't have
+        # to wait a callback from pulseaudio
+        self.default_sink['values'] = list(volume.values)
+        op = lib.pa_context_set_sink_volume_by_index(
+            self.context,
+            self.default_sink['index'],
+            volume,
+            ffi.NULL,
+            ffi.NULL
+        )
+        if op:
+            self.wait_for_operation(op)
+
+    def mute_volume(self):
+        op = lib.pa_context_set_sink_mute_by_index(
+            self.context,
+            self.default_sink['index'],
+            not self.default_sink['muted'],
+            ffi.NULL,
+            ffi.NULL
+        )
+        self.wait_for_operation(op)
+
+    def increase_volume(self, value=2):
+        base = self.default_sink['base_volume']
+        volume = ffi.new('pa_cvolume *', {
+            'channels': self.default_sink['channels'],
+            'values': self.default_sink['values'],
+        })
+        lib.pa_cvolume_inc(
+            volume,
+            int(value * base / 100),
+        )
+        # check that we dont go over 100% in case its set in config
+        if self.limit_max_volume:
+            volume.values = [(i if i <= base else base) for i in volume.values]
```

Comment:
Maybe `max(i, base)`?

Reply:
I think you meant `min(i, base)` and thats the main reason its written verbosely